### DISCUSSION TITLE
is async os.pullEvent() possible?

### DISCUSSION THREAD

**Discussion by C4PTRAMPAGE** [cc-tweaked/CC-Tweaked#1237] [Q&A] [answered]
this problem is the main reason i quit using computercraft out of frustration back in the day, how do i use os.pullEvent() without it stopping the execution of the rest of the script? as far as i can tell parallel is no help here because parallel doesn't work as advertised.

**Comment by Lupus590** [comment]
Due to how ComputerCraft works only a few computers can run at a time. This means that computers have to pause for others to run, Lua does this by yielding, which ComputerCraft wraps as event pulling.

What are you doing that requires async event pulling?

**Reply by C4PTRAMPAGE** [reply]
mostly rednet messages, monitor touch and (via the advanced peripherals chat-box) chat messages, also doing stuff when other addon peripherals fire specific events, which *would* include events from sgcraft but that mod never got updated because the owner has a server and is happy to stay in 1.12 forever and yes i'm more than a little bitter about that. 

anyway the problem is i don't want to use multiple computers to run different parts of a device management script, for example if i were managing a stargate i'd need to listen for rednet messages from the tablets running the gdo (no, thats not a typo) scripts to open the iris while also updating the computercraft monitor with the gates status so i'd end up using two or three computers to do everything even though the scripts they had running are (mostly) small and simple. ...i say that, the script i'm working on now to control stuff in the new tardis mod is over 600 lines and the one back then for the stargate was probably similar.

**Reply by Lupus590** [reply]
Anything that you can do with multiple computers you can probably do with one using parallel. Also, often when parallel is used the code can be rewritten to not use it.

I think this may be the case with your code. Can you upload your code somewhere and I'll try to point you in the right direction?

**Comment by SquidDev** [comment]
> as far as i can tell parallel is no help here because parallel doesn't work as advertised 

This definitely sounds like a documention problem. Would you be able to provide some code you think should work but doesn't?

**Comment by C4PTRAMPAGE** [comment]
the function at line 496 is what i want to run in parallel with everything else.
[https://www.mediafire.com/file/gkes5mspysyss24/TARDOS95.lua/file](url)
and yes i'm using mediafire i cant be bothered to find my pastebin credentials right now.

**Reply by Lupus590** [reply]
https://gist.github.com/ exists and you already are signed in.

**Reply by SquidDev** [reply]
I'd expect `parallel.waitForAll(f_mainCode, f_chatMonitoring)` to work there. Does it not?

**Comment by C4PTRAMPAGE** [comment]
i tried that in the while true loop at the bottom all it did was freeze the main code and only run chat monitoring

**Comment by fatboychummy** [comment, marked as answer]
So you can do this one of two ways. I personally like the second method a bit better.

# Multi-event pulling
Basically, you don't apply a filter to `os.pullEvent()`, then use an `if` statement on the event returned to separate what you want to do.

## Example
```lua
while true do
  local event_data = table.pack(os.pullEvent())
  -- event name is stored as event_data[1]
  
  if event_data[1] == "monitor_touch" then
    -- act on the monitor touch
  
    -- you can do the following if you wish to separate out the values returned by os.pullEvent as well:
    local monitor_touched, x, y = table.unpack(event_data, 2, event_data.n)
    
  elseif event_data[1] == "rednet_message" then
    -- act on the rednet message
  end
end
```

## Cons
1. All your code is in one spot, though you could separate out into functions, ie something like: `monitor_touch(table.unpack(event_data, 2, event_data.n))`
2. If the reaction to an event does anything that requires waiting (like an http request or something), you may miss events.

# Parallel
You can run multiple functions in "parallel." So if you have two functions with a `while true do` loop in them that both yield, you can run them together.

## Example
```lua
local function monitor_handler()
  while true do
    local _, monitor_touched, x, y = os.pullEvent("monitor_touch")
    -- handle the touch event
  end
end

local function rednet_handler()
  while true do
    local _, sender, message, protocol = os.pullEvent("rednet_message")
    -- handle the rednet event
  end
end

parallel.waitForAny(monitor_handler, rednet_handler)
```

## Pros
1. If the monitor is actively doing something, the rednet handler will still receive events.
2. Code is split already, so code that is doing different things are not mixed together.
3. You can choose to either use `parallel.waitForAny` or `parallel.waitForAll`. `waitForAny` will terminate the other handler if one handler stops unexpectedly, whereas `waitForAll` will keep running the other handler until it also stops. This allows some more fine control of how your system will work.

**Reply by C4PTRAMPAGE** [reply]
yeah i don't really get tables at this point... or the file system for that matter not that it applies here, so the second is better.
can't check it now but i'll see if this works later.

**Reply by C4PTRAMPAGE** [reply]
can confirm that doesn't work also i think your not understanding the problem, i'm not trying to listen for different event types at the same time i'm trying to listen for events and run code at the same time, which is to say i want to continue to execute the while true do loop and listen for events at the same time.

i know none of the main code is working because i setup a thing that spins on the screen while the main code is going to visualize the tickrate of the script and when i run the code and listener in parallel the thing doesn't spin.

is there a way to run completely separate scripts on one computer at the same time? that would also work... even thought its still not the no-compromise solution i want.

**Reply by fatboychummy** [reply]
So in one parallel loop just do 

```lua
local function forever()
  while true do
    -- code you want repeated forever
    sleep()
  end
end
```

Then in your other parallel loop, listen for events.

You explicitly MUST yield, which `sleep` does for you. You cannot loop forever without a yield.

And yes, you can also parallelize separate scripts. Either by doing

```lua
parallel.waitForAny(
  function()
    shell.run "script1"
  end,
  function()
    shell.run "script2"
  end
)
```

Or by using the `bg` / `fg` commands on an advanced computer.

```
> bg script1
tab opened
> bg script2
tab opened
> 
```

**Reply by C4PTRAMPAGE** [reply]
ok, i found the problem: i thought that

while true do
  parallel.waitForAll(f_mainCode, f_chatMonitoring)
  sleep(c_tickrate)
end

would be good enough to loop everything but its not, those functions must themselves be while true loops... which it does not say in the documentation. @SquidDev

anyway with that change its all working now, thanks.

**Reply by fatboychummy** [reply]
The documentation states very explicitly exactly what parallel does.

# `waitForAny`
> Switches between execution of the functions, **until any of them finishes.**

# `waitForAll`
> Switches between execution of the functions, **until all of them are finished.**
